### DISCUSSION TITLE
Allow claude to return diffs instead of whole code

### DISCUSSION THREAD

**Discussion by 0n4li** [cline/cline#526] [Feature Requests]
Sometimes, the code changes are minimal. If Claude (or any other model) could return just the diff of the code and at the client side, the full code could be compiled.

This would have below benefits:
1. Faster code generation
2. Fewer output tokens (so less costs)

There could be a tag like <diff></diff> which could be used only when required.

**Comment by NeuralAIM** [comment]
**+1**
_This is the solution to the problem https://github.com/cline/cline/issues/470 and it's the best. Do as in aider, it is much cheaper and works faster._

**Comment by swecjb** [comment]
This has been nicely implemented in the 'Repo Prompt' app, but Cline is better in every other way (terminal, IDE etc.). This is really the only missing feature keeping me from using Cline for all my projects.

**Comment by thebobbyboy420** [comment]
Yeah +1 cline needs to add this feature ASAP

**Comment by devMls** [comment]
+1 for this, it will a big improvement

**Comment by thebobbyboy420** [comment]
This needs fixing ASAP its costing way too much to use Cline compared to Aider

Look how much i spent in a few days crazy 

If Cline needs to edit 1 line of a 500 line file it will write the whole 500 lines just to edit the 1 line. Thats too wasteful

![image](https://github.com/user-attachments/assets/8726e8c1-60eb-4c5c-8aff-0106aff2f31c)

**Reply by justinphilpott** [reply]
So true, its wasteful and so inefficient. "You are missing a brick in your wall? OK! I'll just rebuild the whole wall for you, no problem!!! just hang on a sec..............."

**Comment by M1kep** [comment]
Anthropic provides a "text editor" tool as part of their computer use beta tools, that I think would be perfect for this: https://docs.anthropic.com/en/docs/build-with-claude/computer-use#understand-anthropic-defined-tools

OpenAI has also released Predicted Output: https://platform.openai.com/docs/guides/latency-optimization#use-predicted-outputs for optimizing latency, though for small edits the price would be cheaper, I think the for larger edits it's actually be more expensive.

Though different in nature, I feel like they'd both be huge improvements to the edit flows

**Comment by saoudrizwan** [comment]
I've experimented with structured output like editing certain lines or unified diff format like aider, but I've found the results aren't nearly as good as whole file outputs. The reason is these models are trained on significantly more whole code files than say diff patches, and there's also been studies showing drops in performance whenever you force the models to output in a structured way ie diff or json.

As @M1kep mentioned, OpenAI just released a new API featured called [predicted outputs](https://platform.openai.com/docs/guides/latency-optimization#use-predicted-outputs) that would fix this issue entirely. It essentially spits out the tokens its read before and only has to "think" about new tokens (the new edits to the file), making the edit faster and more reliable without the code truncation. The issue is that this API is very rudimentary and limited (doesn't allow images, only works with gpt4o, etc) so it doesn't fit as nicely into Cline as I'd like. I'm hoping Anthropic comes out with their version of this API soon (which they've pre-released in Zed, called [Fast Edit Mode](https://www.youtube.com/watch?v=6CCuyTMhZKw)), so I can add it to Cline ASAP and fix this problem.

Pls join the [discord](https://discord.gg/cline) for updates about this, and thank you for bearing with me in the meantime 🙏

**Reply by GraydonCompanies** [reply]
>>>I've experimented with structured output like editing certain lines or unified diff format like aider, but I've found the results aren't nearly as good as whole file outputs.

Fair enough, but wouldn't it be better to give users the option as long as they are aware that it doesn't work well in certain cases? Cline is absolutely the best VS Code AI extension, but the cost is just unsustainable.

**Reply by drcggray** [reply]
Thanks for your amazing work creating this incredible tool @saoudrizwan !!! 😁

**Reply by MusiCode1** [reply]
I think if you let the model write whole functions instead of whole files, we'll get the best result.
While the model isn't good at writing diff code, it does write whole functions well.
And writing a whole function will be shorter than writing a whole file with several functions.

Claude Sonnet has a nasty habit of writing placeholders like "the rest of the code is the same", instead of writing the full code.
What if we could get it to write instead of a placeholder like "the rest of the code is the same", so that it would write "lines 50 to 68 are the same" in output that the code editor can understand?
That would solve all the problems at once!

**Reply by yisrael-rosen** [reply]
The core problem with using line numbers is that language models don't actually "think" in terms of line numbers. They process text as token sequences without maintaining a strict line-by-line understanding that would match an editor's view.

In my [fork](https://github.com/yisrael-rosen/cline-go), I tried creating a tool that replaces functions based on function names. Using VSCode's generic API 'vscode.executeDocumentSymbolProvider', it didn't always parse the code correctly and needed to fall back to regular writing.

Since I mainly write in Go, I added a Go CLI tool for analyzing Go code, which performed much better. I need to investigate if this approach would be efficient for other languages, especially OOP languages where code structure is more complex. Also, I need to consider how to handle code comments that exist outside of functions.

**Comment by KartDriver** [comment]
I think if you prompt the model properly, it can be done very well.  I've had decent look working on things at the function level and just asking for the full functions that are changed or added.  You really don't even need to have the model do the patch formatting, you could do that client side if it impacts model performance.

I think it would be handy to make it optional, so users could continue to use the full method if they prefer, or select the partial updates.

**Comment by eoakley** [comment]
This is Cursor's approach, it is very simple, but might require a smart "Apply" function or model to work properly.

```
You are happy to help answer any questions that the user has (usually they will be about coding).

1. When the user is asking for edits to their code, please output a simplified version of the code block that highlights the changes necessary and adds comments to indicate where unchanged code has been skipped. For example:
language:path/to/file
// ... existing code ...
{{ edit_1 }}
// ... existing code ...
{{ edit_2 }}
// ... existing code ...

The user can see the entire file, so they prefer to only read the updates to the code. Often this will mean that the start/end of the file will be skipped, but that's okay! Rewrite the entire file only if specifically requested. Always provide a brief explanation of the updates, unless the user specifically requests only the code.

2. Do not lie or make up facts.

```

**Reply by fiveze** [reply]
Where can I read about this approach?

**Comment by KartDriver** [comment]
The cursor approach is interesting.  It's certainly more efficient than the search/replace format of aider, since that still requires outputting chunks of old code, but it would take a lot more code and effort to handle the replacement/diff functionality.

It would definitely be helpful to have any sort of partial updates option.  Right now, if your source code files are too big, you just can't work on them with Cline, not to mention the costs factor.  I usually ask for the just the code changes  to be output to chat when this happens and then I either manually copy/paste it, or *gasp* I fire up Cursor and ask it to make the changes.

**Comment by thunder-007** [comment]
@saoudrizwan How about a toggle which will let the user choose to use replace only necessary or whole

**Comment by asbis** [comment]
I had good experience with this change:
https://github.com/asbis/autocoder/blob/automerger/src/core/prompts/system.ts
It works best with sonnet of course.


You can try out my branch:
https://github.com/asbis/autocoder

It lets you run "auto-accept" and mostly it can work by itself without too much help. But still working on it to make it work perfect.
I can sometimes just leave it for 1 hour and it have made a simple functional app.

**Comment by saoudrizwan** [comment]
### Cline v3 now supports diff editing!

When Cline needs to make targeted edits to large files, he now uses a search & replace diff format where he only has to output the content to replace, rather than the whole file. This allows for faster, more reliable edits and prevents code deletion issues where previously, unmodified sections of code might have been replaced with comments like `// rest of code here`. Cline can also fall back to whole file editing for when it makes sense, e.g. replacing boilerplate files, complete refactors, or when diff editing fails.

**Reply by GraydonCompanies** [reply]
Thank you, Saoud! You're the best and Cline is amazing!

**Comment by 0n4li** [comment]
Yes it works. Models need to get better at returning diffs as I see a lot of errors with Gemini 1206 and even Sonnet 3.6.
However, this is a great start. Thanks!
